src/java/org/apache/cassandra/cql3/Sets.java

```diff
@@ -348,26 +349,40 @@ public void execute(DecoratedKey partitionKey, UpdateParameters params) throws I
 
         static void doAdd(Term.Terminal value, ColumnMetadata column, UpdateParameters params) throws InvalidRequestException
         {
+            if (value == null)
+            {
+                // for frozen sets, we're overwriting the whole cell
+                if (!column.type.isMultiCell())
+                    params.addTombstone(column);
+
+                return;
+            }
+
+            SortedSet<ByteBuffer> elements = ((Value) value).elements;
+
             if (column.type.isMultiCell())
             {
-                if (value == null)
-                    return;
+                // Guardrails about collection size are only checked for the added elements without considering
+                // already existent elements. This is done so to avoid read-before-write, having additional checks
+                // during SSTable write.
+                Guardrails.itemsPerCollection.guard(elements.size(), column.name.toString(), params.clientState);
 
-                for (ByteBuffer bb : ((Value) value).elements)
+                int dataSize = 0;
+                for (ByteBuffer bb : elements)
```

Comment:
Why don't we have this `UNSET_BYTE_BUFFER` check on `Lists` i.e.?

Reply:
I think that unset values aren't supported for lists and maps, that was done in CASSANDRA-7304
 and this ticket doesn't change it, maybe we should reconsider on a followup ticket?

Reply:
GTK about c7304. In any case definitely outside the scope of this ticket.